pkg/util/inplaceupdate/inplace_update_defaults.go

```diff
@@ -487,3 +511,391 @@ func checkAllContainersHashConsistent(pod *v1.Pod, runtimeContainerMetaSet *apps
+				containersResourceChanged.Insert(c.Name)
+			}
+		} else {
+			state.NextContainerImages[c.Name] = newImage
+			state.NextContainerResources[c.Name] = newResource
+		}
+	}
+
+	// This provides a hook for vertical updates,
+	// so that internal enterprise implementations can update pod resources here at once
+	verticalUpdateOperator.UpdatePodResource(pod)
+
+	for _, c := range pod.Status.ContainerStatuses {
+		if containersImageChanged.Has(c.Name) {
+			if state.LastContainerStatuses == nil {
+				state.LastContainerStatuses = map[string]appspub.InPlaceUpdateContainerStatus{}
+			}
+			if cs, ok := state.LastContainerStatuses[c.Name]; !ok {
+				state.LastContainerStatuses[c.Name] = appspub.InPlaceUpdateContainerStatus{ImageID: c.ImageID}
+			} else {
+				cs.ImageID = c.ImageID
+			}
+		}
+		if containersResourceChanged.Has(c.Name) {
+			verticalUpdateOperator.SyncContainerResource(&c, state)
+		}
+	}
+
+	// This provides a hook for vertical updates,
+	// so that internal enterprise implementations can sync pod resources here at once
+	verticalUpdateOperator.SyncPodResource(pod, state)
+
+	// update annotations and labels for the containers to update
+	for cName, objMeta := range spec.ContainerRefMetadata {
+		if containersToUpdate.Has(cName) {
+			for k, v := range objMeta.Labels {
+				pod.Labels[k] = v
+			}
+			for k, v := range objMeta.Annotations {
+				pod.Annotations[k] = v
+			}
+		} else {
+			state.NextContainerRefMetadata[cName] = objMeta
+		}
+	}
+
+	// add the containers that update this time into PreCheckBeforeNext, so that next containers can only
+	// start to update when these containers have updated ready
+	// TODO: currently we only support ContainersRequiredReady, not sure if we have to add ContainersPreferredReady in future
+	if len(state.NextContainerImages) > 0 || len(state.NextContainerRefMetadata) > 0 || len(state.NextContainerResources) > 0 {
+		state.PreCheckBeforeNext = &appspub.InPlaceUpdatePreCheckBeforeNext{ContainersRequiredReady: containersToUpdate.List()}
+	} else {
+		state.PreCheckBeforeNext = nil
+	}
+
+	state.ContainerBatchesRecord = append(state.ContainerBatchesRecord, appspub.InPlaceUpdateContainerBatch{
+		Timestamp:  metav1.NewTime(Clock.Now()),
+		Containers: containersToUpdate.List(),
+	})
+
+	klog.V(5).Infof("Decide to in-place update pod %s/%s with state %v", pod.Namespace, pod.Name, util.DumpJSON(state))
+
+	inPlaceUpdateStateJSON, _ := json.Marshal(state)
+	pod.Annotations[appspub.InPlaceUpdateStateKey] = string(inPlaceUpdateStateJSON)
+	return pod, nil
+}
+
+// defaultCalculateInPlaceUpdateSpec calculates diff between old and update revisions.
+// If the diff just contains replace operation of spec.containers[x].image, it will returns an UpdateSpec.
+// Otherwise, it returns nil which means can not use in-place update.
+func defaultCalculateInPlaceUpdateSpecWithVerticalUpdate(oldRevision, newRevision *apps.ControllerRevision, opts *UpdateOptions) *UpdateSpec {
```

Comment:
I am inclined to abstract the core logic function and call it within defaultCalculateInPlaceUpdateSpec based on feature gate.

Or abstract the calcOption in defaultCalculateInPlaceUpdateSpec and use calcOption to assemble.

The reason is: this feature is an enhancement to the current implementation, not a fork. The current way of implementation has too much repetitive code.